rlclientlib/utility/context_helper.cc

```diff
@@ -1,65 +1,55 @@
-#include <cpprest/json.h>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <rapidjson/document.h>
+#include <rapidjson/prettywriter.h>
 #include <object_factory.h>
-#include <cpprest/asyncrt_utils.h>
 #include "err_constants.h"
-#include "http_helper.h"
 
 #include <chrono>
 #include <cstring>
 
-namespace sutil = ::utility::conversions;
+using namespace rapidjson;
 
 namespace reinforcement_learning { namespace utility {
-  const auto multi = sutil::to_string_t("_multi");
-  const auto slots = sutil::to_string_t("_slots");
-  const auto event_id = sutil::to_string_t("_id");
+  const auto multi = "_multi";
+  const auto slots = "_slots";
+  const auto event_id = "_id";
+
   /**
-   * \brief Get the number of actions found in context json string.  Actions should be in an array
-   * called _multi under the root name space.  {_multi:[{"name1":"val1"},{"name1":"val1"}]}
-   *
-   * \param count   : Return value passed in as a reference.
-   * \param context : String with context json
-   * \param status  : Pointer to api_status object that contains an error code and error description in
-   *                  case of failure
+   * \brief Get the event IDs from the slots entries in the context json string.
+   * 
+   * \param context   : String with context json
+   * \param event_ids : Reference to the mapping from slot index to event ID string, results will be
+   *                    put in here
+   * \param trace     : Pointer to the trace logger
+   * \param status    : Pointer to api_status object that contains an error code and error description in
+   *                    case of failure
    * \return  error_code::success if there are no errors.  If there are errors then the error code is
    *          returned.
    */
-  int get_action_count(size_t& count, const char *context, i_trace* trace, api_status* status) {
+  int get_event_ids(const char* context, std::map<size_t, std::string>& event_ids, i_trace* trace, api_status* status) {

```

Comment:
Do you need the keys to be in sorted order ? If not, I would recommend to use std::unordered_map() which give O(constant) for insertion and check.

Reply:
Hmm, I think that would be fine, but it was using std::map before, as well. @lokitoth would this be a valid optimization?

Reply:
At a guess, based on where this code is being used: The expectation is that the number of entries to this map will be small (as it is capped by the number of slots in a given CCB request), and immediately after getting out of this call, it gets iterated, start-to-finish.

(A better solution here might be to turn it into a "foreach_event_id" template, to avoid constructing this map in the first place, but that is out of the scope of this change, I think.)

cc @jackgerrits who may have more context about the history here

Reply:
Yeah, I am fine with keeping it std::map, just wanted to bring it up because of potential runtime complexity improvement here.